Tried an approch for text case in pixelmatch

Vaivaswat2244 opened the pull request:
<!--
  Thank you for contributing! Please use this pull request (PR) template.


 In the description field of this PR, include "resolves #XXXX" tagging the issue you are fixing. If this PR addresses the issue but doesn't completely resolve it (ie the issue should remain open after your PR is merged), write "addresses #XXXX".-->
Resolves #7496

 Changes:
<!-- Add here what changes were made in this pull request and if possible provide links showcasing the changes. -->
Modified the `checkMatch` function in `visualtest.js` to be more tolerant of slight text position shifts while still catching meaningful changes. Key changes:

Added text detection using pixel pattern analysis:

- Looks for moderate density and high contrast transitions typical of text
- Helps identify regions where small shifts should be allowed


Implemented shift tolerance for text regions:

- Allows up to 2-pixel shifts for detected text areas
- Still fails on actual content changes
- Configurable via `shiftThreshold` parameter

 Screenshots of the change:
<!-- If applicable, add screenshots depicting the changes. -->

#### PR Checklist
<!--
  To check any option, replace the "[ ]" with a "[x]". Be sure to check out how it looks in the Preview tab! Feel free to remove any portion of the template that is not relevant for your issue.
-->

- [x] `npm run lint` passes
- [ ] [Inline reference] is included / updated
- [x] [Unit tests] are included / updated

[Inline reference]: https://p5js.org/contribute/contributing_to_the_p5js_reference/
[Unit tests]: https://github.com/processing/p5.js/tree/main/contributor_docs#unit-tests

davepagurek commented:
Thanks @Vaivaswat2244, nice approach! I have a few questions that might help us determine how to get this across the finish line.

- Do you have a sense of how often `isLikelyText` returns true, and if this gets called on anything non-text? I wonder if maybe rather than trying to detect whether we think there is text, we can directly record when `text` is called. Since we pass in a p5 instance to the test case anyway, maybe we can do something like this to patch `text()` to record that it got called:

  ```js
  let hasText = false
  const prevText = myp5.text
  myp5.text = function(...args) {
    hasText = true
    prevText.apply(this, args)
  }
  ```
  
 - It looks like [some tests are failing on CI now.](https://github.com/processing/p5.js/actions/runs/13353056666/job/37292021733?pr=7552) Do you know if these look like (1) false positives that the checker needs to address, (2) cases where we need to regenerate the screenshots because the old ones were out of date but still passed on CI under the old pixel checker, or (3) are actual bugs in p5 that are surfaced now by using a better pixel checker?

Vaivaswat2244 commented:
yeah sure, the text() passing makes much more sense. I can implement that.
And regarding the tests getting failed. I think that these are the 2nd case as you described, failed the CI tests after regenerating on my machine but were passed by the old checker. Also I'm not so sure how to move ahead from here. So ig for now I'll change the text case and will see how that works out.

davepagurek commented:
Are you able to see the test output from CI on your PR by clicking the Details button here?
<img width="756" alt="image" src="https://github.com/user-attachments/assets/d5042509-58bf-4e48-8c19-3f503cbb6b17" />

For a next step, I'd look at the test output, and try regenerating the failed screenshots locally and pushing those. We can see if that fixes it on CI, and also by looking at the new images in the Files Changed tab, whether we think the changed images are expected.

Vaivaswat2244 commented:
hey, so I just noticed that the files which I added in the earlier commit aren't the one which are failing the tests on CI, but they are other files which were not touched.

Now, I changed the files which were failing tests in CI, and tried the tests locally. The tests are now failing for different files in the typography and webgl cases. I guess there is a problem with my changes in the visualtest file.
So maybe I'll revert these changes and think of something else.

davepagurek commented:
Just looking at the last failure in the test file:

Expected:
![image](https://github.com/user-attachments/assets/4365a056-a679-43fa-bbd9-30f0c0998f73)

Received:
![image](https://github.com/user-attachments/assets/ae05b4c9-96d7-4faf-bad9-05d24ec56126)

Diff:
![image](https://github.com/user-attachments/assets/5dc95eeb-9cc5-4df6-aee9-14b181e384a6)

...it looks like a few isolated pixels are off. I wonder if we can do something to augment pixelmatch and check if the difference exists just on a single pixel (so no surrounding pixels are also flagged) and then ignore those?

Vaivaswat2244 commented:
Sure, I can try that

davepagurek commented:
Nice, the clustering idea is good and gives us some options for dealing with issues like this! I suspect if a shape is consistently 1px thick, then it could be the result of a shift due to environment differences. 

A way to check that might be to try to see if, for every pixel in the cluster, it has at most two neighbours?

Vaivaswat2244 commented:
I have made some changes to this, so now additional to clustering logic our algo does the following
- Iterates through all 8 neighboring pixels (excluding itself) to determine how many are part of the difference.
- If more than 80% of the pixels in the cluster have ≤2 neighbors, the difference is classified as a line shift (isLineShift = true).

Using this and keeping the thresholds same as the more tolerable [commit](https://github.com/processing/p5.js/pull/7552/commits/8f50108ca23b05e38ee3647f9c3d05d05fa9738d), the tests are now passing.
After a game of trial and error in  the thresholds, found that 0.5 for pixelmatch is optimum for us, as this is passing the tests in default cases. And in the pr #7495 , I reversed the changes and added pervVertex and the test which you added failed. 

Expected:
![image](https://github.com/user-attachments/assets/e2865b3e-0c5f-464f-adbc-53d5ec7acb57)

Recieved:
![image](https://github.com/user-attachments/assets/0cb1b4f5-4e8b-409e-ac95-ec8146149687)

Diff: 
![image](https://github.com/user-attachments/assets/2b4371ca-1ab1-476a-8d20-21a3fafa7b27)

So is there anything else I should change.

davepagurek commented on test/unit/visual/visualTest.js:
```diff
@@ -88,55 +89,213 @@ export function visualSuite(
+  // Define significance thresholds
+  const MIN_CLUSTER_SIZE = 4;  // Minimum pixels in a significant cluster
+  const MAX_TOTAL_DIFF_PIXELS = 40;  // Maximum total different pixels
+  const MAX_LINE_SHIFT_PIXELS = 200;
```
Does this get used?

Vaivaswat2244 replied:
Yeah these are getting used in the check match function itself for final comparison. 
Variation of these thresholds also causes variations in the tests results as well, like the Min_Cluster_Size if reduced from this makes our test less tolerable and can cause false CI fails like we saw in the isolated pixel case.

davepagurek reviewed (commented):
Awesome, this is looking really promising, great work! I think the algorithm is in a spot where it's good to go. The last thing might be to add a comment somewhere explaining the general approach of the diff algorithm, similar to what you've described in comments to me here, and maybe additionally why it's necessary at all (because contributors running tests just on their own system might not realize that there will be differences on CI.)

Vaivaswat2244 commented on test/unit/visual/visualTest.js:
```diff
@@ -88,55 +89,213 @@ export function visualSuite(
     scale *= 2;
   }
-
+  
```
Right!
